File: mmcv/ops/all_pairs_correlation.py
# Copyright (c) OpenMMLab. All rights reserved.
import torch
from torch import Tensor, nn
from torch.autograd import Function
from torch.autograd.function import once_differentiable

from ..utils import ext_loader

ext_module = ext_loader.load_ext(
    '_ext',
    ['all_pairs_correlation_forward', 'all_pairs_correlation_backward'])


class AllPairsCorrelationFunction(Function):

    @staticmethod
    def forward(ctx, input1, input2):
        ctx.save_for_backward(input1, input2)
        iH1, iW1 = input1.shape[-2:]
        iH2, iW2 = input2.shape[-2:]
        batch_size = input1.shape[0]
        output_size = (batch_size, iH1, iW1, iH2, iW2)

        output = input1.new_empty(output_size)

        ext_module.all_pairs_correlation_forward(input1, input2, output)
        return output

    @staticmethod
    @once_differentiable
    def backward(ctx, grad_output):
        input1, input2 = ctx.saved_tensors
        grad_input1 = torch.zeros_like(input1)
        grad_input2 = torch.zeros_like(input2)

        ext_module.all_pairs_correlation_backward(grad_output, input1, input2,
                                                  grad_input1, grad_input2)

        return grad_input1, grad_input2


class AllPairsCorrelation(nn.Module):
    """All-pairs correlation operator.

    All-pairs correlation is used to compute visual similarity in `RAFT
        <https://link.springer.com/chapter/10.1007/978-3-030-58536-5_24>`_.
    The correlation output with the shape (N, H, W, H, W) is formed by taking
    the dot product between all pairs of input feature vectors , and H, W is
    equal the input tensors'.
    """

    def __init__(self) -> None:
        super().__init__()

    def forward(self, input1: Tensor, input2: Tensor) -> Tensor:
        return AllPairsCorrelationFunction.apply(input1, input2)

    def __repr__(self) -> str:
        s = self.__class__.__name__
        return s

File: mmcv/ops/csrc/pytorch/all_pairs_correlation.cpp
// Copyright (c) OpenMMLab. All rights reserved.
#include <iostream>
#include "pytorch_cuda_helper.hpp"
#include "pytorch_cpp_helper.hpp"

#ifdef MMCV_WITH_CUDA

void AllPairsCorrelationForwardCUDAKernelLauncher(Tensor input1, Tensor input2,
                                                  Tensor output);

void AllPairsCorrelationBackwardCUDAKernelLauncher(Tensor grad_output,
                                                   Tensor input1, Tensor input2,
                                                   Tensor grad_input1,
                                                   Tensor grad_intput2);
void all_pairs_correlation_cuda_forward(Tensor input1, Tensor input2,
                                        Tensor output)
{
    AllPairsCorrelationForwardCUDAKernelLauncher(input1, input2, output);
}

void all_pairs_correlation_cuda_backward(Tensor grad_output,
                                         Tensor input1, Tensor input2,
                                         Tensor grad_input1,
                                         Tensor grad_intput2)
{
    AllPairsCorrelationBackwardCUDAKernelLauncher(grad_output,
                                                  input1, input2,
                                                  grad_input1,
                                                  grad_intput2);
}
#endif

void all_pairs_correlation_forward(Tensor input1, Tensor input2, Tensor output)
{
    if (input1.device().is_cuda() and input2.device().is_cuda())
    {
#ifdef MMCV_WITH_CUDA
        CHECK_CUDA_INPUT(input1);
        CHECK_CUDA_INPUT(input2);
        all_pairs_correlation_cuda_forward(input1, input2, output);
#else
        AT_ERROR("Correlation is not compiled with GPU support");
#endif
    }
    else
    {
        AT_ERROR("All-pairs correlation is not implemented on CPU");
    }
}

void all_pairs_correlation_backward(Tensor grad_output,
                                    Tensor input1, Tensor input2,
                                    Tensor grad_input1, Tensor grad_input2)
{
    if (input1.device().is_cuda() and input2.device().is_cuda())
    {
#ifdef MMCV_WITH_CUDA
        CHECK_CUDA_INPUT(grad_output);
        CHECK_CUDA_INPUT(input1);
        CHECK_CUDA_INPUT(input2);
        all_pairs_correlation_cuda_backward(grad_output, input1, input2,
                                            grad_input1, grad_input2);

#else
        AT_ERROR("All-pairs correlation is not compiled with GPU support");
#endif
    }
    else
    {
        AT_ERROR("Correlation is not implemented on CPU");
    }
}
